full screen exclusive

gpx1000 opened the pull request:
## Description

Add's an example of Full Screen Exclusive.  This is a Windows only sample as it's a Windows only problem that is being solved.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

wflohry committed:
Using basic raytracing as template

wflohry committed:
Load sponza

wflohry committed:
Loading sponza model

wflohry committed:
Separating out ray tracing command buffers

wflohry committed:
Add render modes

wflohry committed:
Edit vertex structure

wflohry committed:
Put static objects in buffer with device-exclusive memory

wflohry committed:
Adding vertex and index data to bindings

wflohry committed:
Demonstration of referencing model data from ray tracing shader

wflohry committed:
Implemented texture array and visualization

wflohry committed:
Added fire pit

wflohry committed:
Added ray-traced shadows

wflohry committed:
Adding support for AO

wflohry committed:
Updated model settings

gpx1000 committed:
Fixed a few warnings and spelling errors.

Also fixed the fences problem that causes crash bug from fences on close.

wflohry committed:
Separating out static model loading

wflohry committed:
Separate handling of static / dynamic buffers

wflohry committed:
Added dynamic object and refactored

gpx1000 committed:
Fixed another warning and fixed the fences.

gpx1000 committed:
Merge branch 'feature/raytrace-extended' into warn_removal_and_fence_fix

# Conflicts:
#	samples/extensions/raytracing_extended/raytracing_extended.cpp

gpx1000 committed:
Switch to using constants and specializations in order to change values.  Can switch back to uniforms at a later date but the uniforms were proving a render issue for linux.  This should also demonstrate how to use multiple specialization constants in a ray traced environment.

gpx1000 committed:
more warnings removal

wflohry committed:
Merge warn_remove_and_fence_fix

gpx1000 committed:
Merged in warn_removal_and_fence_fix (pull request #1)

Fixed a few warnings and spelling errors.

Approved-by: William Lohry

wflohry committed:
Merge branch 'feature/raytrace-extended' of ssh://bitbucket.org/holochip/vulkan-samples into feature/raytrace-extended

wflohry committed:
Moving instance buffer creation to TLAS construction

wflohry committed:
Adding refraction placeholder

wflohry committed:
Perform BLAS update for dynamic objects and TLAS update every frame.

wflohry committed:
Adding dynamic sinusoidal wave

wflohry committed:
Finished refractive wave

wflohry committed:
Update to refraction

wflohry committed:
Adding separate flame instance

wflohry committed:
Working flame / particle generator

wflohry committed:
Updates to flame and refraction calculation

wflohry committed:
Updated refraction algorithm

wflohry committed:
Added texture

gpx1000 committed:
more warnings removal
move the png to ktx format

gpx1000 committed:
ensure we're pointing to the right place

wflohry committed:
Merge and update

wflohry committed:
Update to ktx commit

gpx1000 committed:
run clang-format over the sample.

wflohry committed:
Added transformations

wflohry committed:
Merged

wflohry committed:
Merged

gpx1000 committed:
Merged in clang-format (pull request #3)

run clang-format over the sample.

Approved-by: William Lohry

wflohry committed:
Adding transformations

wflohry committed:
Implemented automatic  billboard rotation

wflohry committed:
Merge branch 'feature/raytrace-extended' of ssh://bitbucket.org/holochip/vulkan-samples into feature/raytrace-extended

gpx1000 committed:
run a profile pass.
move pts, uvs, normals, indices to be created/initialized once to avoid dynamic memory generation.
remove warnings.
Update copyright.

gpx1000 committed:
Merged in profile (pull request #4)

run a profile pass.

Approved-by: William Lohry

gpx1000 committed:
Update copyright in CMakeLists.txt
Update sample cmake function definition
Update samples list in Samples CMakeLists.txt
Update README.md with first pass definition of the sample.

gpx1000 committed:
Merged in draft_ready (pull request #5)

Update copyright in CMakeLists.txt

Approved-by: William Lohry

wflohry committed:
Added tutorial rough draft

gpx1000 committed:
grammar check.

gpx1000 committed:
grammar check.

gpx1000 committed:
Merged in tutorial (pull request #6)

grammar check.

Approved-by: William Lohry

wflohry committed:
Added fence reset

wflohry committed:
Updating triangle indexing code

wflohry committed:
Avoid dynamic allocation / copy

gpx1000 committed:
update copyright

gpx1000 committed:
Merged in feature/raytrace-extended (pull request #7)

Feature/raytrace extended

gpx1000 committed:
Merge branch 'downstream-master'

gpx1000 committed:
Fixes for Clang-format and copyright headers CI failures.

gpx1000 committed:
add in nonuniformEXT to the texture lookup as imageOffset isn't uniform.

gpx1000 committed:
switch to textureLod from texture to ensure it's ready for mipmaps if copied from sample.

wflohry committed:
Fixing refraction bug

gpx1000 committed:
Merge remote-tracking branch 'downstream/feature/raytrace-extended'

gpx1000 committed:
fix typo in README.md

gpx1000 committed:
fix typo in README.md and in CMakeLists.txt

gpx1000 committed:
Merge branch 'master' into master

wflohry committed:
Updates for PR comments

wflohry committed:
Adding memory barriers

wflohry committed:
Updating shaders to use defines

gpx1000 committed:
Merge remote-tracking branch 'downstream/feature/raytrace-extended'

# Conflicts:
#	samples/extensions/raytracing_extended/raytracing_extended.cpp
#	shaders/khr_ray_tracing_extended/closesthit.rchit

gpx1000 committed:
clang format cleanup

gpx1000 committed:
add missing copyright to new header

gpx1000 committed:
fix attribution in the tutorial.md

gpx1000 committed:
framework:
acceleration structure: (build) fix a bug which only would allow a single triangle to be included in a acceleration structure.
acceleration structure: (build) allow for more than one call to build (useful in animation).  This was caused by recreating the buffer each time rather than reusing it if it already exists.
acceleration structure: (add triangles) allow for developer to specify the data address of the vertex index and transform buffers that are used.  This allows for multi model gltf files to work as in the sponza scene.
acceleration structure: enable buffer retrieval to allow for developer to add in memory barriers.
acceleration structure: provide developer means to wipe the geometries added to the acceleration structure.  This is very important if animating and thus recalling the same structure(s).  NB: a future feature would be to index the geometries and instances to allow for updating existing data individually rather than a requirement to wipe out everything and rebuild everything.
scratch buffer: add a return for the scratch buffer's size to allow for reusing the scratch buffer if it's the same size as previously allocated.

Sample:
As this is the first ray tracing sample to make use of @Sascha William's excellent acceleration structure framework, and we already have the existing long form, I opted to enable a macro to turn on/off using the framework's acceleration structure.  I think this is valuable to end users to show a precise "what the frameork does" behind the scenes.  Future samples should opt to use the framework.

gpx1000 committed:
run clang-format over scratch_buffer.h

gpx1000 committed:
Add in proper update logic for the acceleration_structure in both the framework and the sample.

gpx1000 committed:
Ensure that only updated geometry are used in each update pass and make framework the default AS.

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Small changes for buffer.

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
run clang-format

gpx1000 committed:
fix android

gpx1000 committed:
fix android

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
address the returns doc.

gpx1000 committed:
Merge branch 'KhronosGroup:master' into master

gpx1000 committed:
Merge branch 'upstream-master'

gpx1000 committed:
Merge branch 'KhronosGroup:master' into master

gpx1000 committed:
Merge branch 'upstream-master'

gpx1000 committed:
Merge branch 'upstream-master'

gpx1000 committed:
Merge branch 'KhronosGroup:master' into master

gpx1000 committed:
Merge remote-tracking branch 'origin/master'

gpx1000 committed:
Fix issue #452

gpx1000 committed:
Merge branch 'upstream-master'

JeremyG2101 committed:
PLEASE DO NOT MERGE IT! This is a Debug sample including: memory_budget

JeremyG2101 committed:
working version update

JeremyG2101 committed:
Quick fix just removed 2 test log lines

JeremyG2101 committed:
All changes were made based on Steve's comments. Syntax warnings were claned from my end. All printf related functions/commands were removed.

JeremyG2101 committed:
copyrights year should be 2019 based on the instancing created

JeremyG2101 committed:
CMake list update.

JeremyG2101 committed:
add copyrights to CMake List

JeremyG2101 committed:
Per Steve's Comments: 1) In the CMakeList, the introduction is now 100% focus on what was introduced related to the memory budget extension feature. 2)update_memory_properties() function is now REMOVED from the draw() and reintroduced to the prepare_instance_data() function, after everything is calculated, loaded, and staged. 3) All runtime update/button-press-update related boolean and functions were REMOVED.

JeremyG2101 committed:
The Tutorial is fianlly done, and the image is saved in .PNG format, organized in an individual folder since the memory_budget folder. the tutorial is named as README.md

JeremyG2101 committed:
This update regards to Steve's newest comments to revise the description modification copyrights to Holochip Corporation; and insert additional description of the usage of update_device_memory_properties().

JeremyG2101 committed:
Quick syntax fix on "the" to "The" from description of int variable "mesh_density".

JeremyG2101 committed:
went thfough the clang check and already have the format modified, our project name is added to the CmakeList

JeremyG2101 committed:
Merge pull request #8 in holochip/vulkan-samples from Memory_Budget_Extension to master

Memory_Budget_Debug_Sample

Approved by Steven Winston

JeremyG2101 committed:
FR-406: Memory Budget: Quick fix corresponding for the GitHub comments regarding memory units conversion issues and get_memory_properties2 call inside prepare()

JeremyG2101 committed:
Code revised per Steve's new comments 10/25/2022

JeremyG2101 committed:
This is the dummy renderer class built for full_screen_exclusive extension sample, after it passed check, I will create an individual branch just for a dummy renderer, it I don't have to write 1,000+ lines to begin with. Where I can directly dive into add models and change skybox.

JeremyG2101 committed:
So this is the clean and neat version of it, Clang check applied.

JeremyG2101 committed:
code revised and set private to extension related variables

JeremyG2101 committed:
Full Screen Extension exclusive

JeremyG2101 committed:
Organized the full Screen Exclusive branch, which it is actually corresponding to the current/up-to-date changes of the memory_budget. It also adds the full_screen_exclusive to the project CMake list.

JeremyG2101 committed:
Adds more annotations for display mode selection

JeremyG2101 committed:
noticed an incorrect annotation

gpx1000 committed:
Merge branch 'KhronosGroup:master' into master

JeremyG2101 committed:
A clean version of this

gpx1000 committed:
Merged master into Full_Screen_Exclusive_Extension

gpx1000 committed:
Merge branch 'upstream-master'

gpx1000 committed:
Merge remote-tracking branch 'origin/master'

gpx1000 committed:
Merge branch 'master' into downstream-master

# Conflicts:
#	samples/CMakeLists.txt

gpx1000 committed:
Revert memory budget should NOT be merged to master.

gpx1000 committed:
Revert "This update regards to Steve's newest comments to revise the description modification copyrights to Holochip Corporation; and insert additional description of the usage of update_device_memory_properties()."

This reverts commit 889ece87b111e808e71a4d98069fe7df1ccc0b68.

gpx1000 committed:
Revert "Quick syntax fix on "the" to "The" from description of int variable "mesh_density"."

This reverts commit 35e7dae9ff8a859a1905867236dd88fc6cd22fd2.

gpx1000 committed:
Revert "The Tutorial is fianlly done, and the image is saved in .PNG format, organized in an individual folder since the memory_budget folder. the tutorial is named as README.md"

This reverts commit 3cc5023045cf19ee3ad1e011f35ae84b394d9597.

gpx1000 committed:
Revert "Per Steve's Comments: 1) In the CMakeList, the introduction is now 100% focus on what was introduced related to the memory budget extension feature. 2)update_memory_properties() function is now REMOVED from the draw() and reintroduced to the prepare_instance_data() function, after everything is calculated, loaded, and staged. 3) All runtime update/button-press-update related boolean and functions were REMOVED."

This reverts commit 8cdc982c22e040ace07e62a866758e3e58e99a81.

gpx1000 committed:
Revert "copyrights year should be 2019 based on the instancing created"

This reverts commit 428cc6eab40f504a1b846995f1e9274051daae62.

gpx1000 committed:
Revert "add copyrights to CMake List"

This reverts commit 1086cab586b2dfa78eac2b111c5ff4d39779fec9.

gpx1000 committed:
Revert "CMake list update."

This reverts commit 8c14c62074c1bb713e35da3f546a6a5009de9a90.

gpx1000 committed:
Revert "All changes were made based on Steve's comments. Syntax warnings were claned from my end. All printf related functions/commands were removed."

This reverts commit 8970f8fdce9eaa76d743748982be4d06e6a5be61.

gpx1000 committed:
Revert "Quick fix just removed 2 test log lines"

This reverts commit 7708fff86dd2adf80aa37efb9e122b117c18cd7d.

gpx1000 committed:
Revert "working version update"

This reverts commit 9a56f3b7bb387f683d6b08bb89becb1110e7256d.

gpx1000 committed:
Revert "PLEASE DO NOT MERGE IT! This is a Debug sample including: memory_budget"

This reverts commit f6d3730ebcfe9a4f96451133e18fddf8b3af03ef.

gpx1000 committed:
Merge branch 'downstream-master' into Full_Screen_Exclusive_Extension

gpx1000 committed:
update version of vulkan the submodules are using.

gpx1000 committed:
update version of vulkan the submodules are using.

JeremyG2101 committed:
added all fixes according to the last commonts. This is also a quick save point before I add the customized swapchain, swapchain image, and image view functions. The mentioned function set will ultimately replaces the swapchain and image views created from the render_context class, but remains everything else the same.

JeremyG2101 committed:
full screen exclusive are updated based on framework changes made in the most up-to-date vulkan sample version. In addition to which, 3 bugs are fixed (suggested to fix) in response to the newest version of glslang, regarding to the directory/directories change(s) made to LimitedResources.

JeremyG2101 committed:
This update adds an entire image view recreate function in addition to the swapchain recreation. Meantime, the swapchain images are are deep copied to a new image and sync them after the swapchain recreation. Clang checked for both files.

JeremyG2101 committed:
Add a boolean to activate or deactive swapchain and full screen exclusive related functions

JeremyG2101 committed:
This build uses Application as the base class of its build due to inaccessiblity to the "swapchain" class and its corresponding images, image-views (swapchain_buffers).

JeremyG2101 committed:
This adds the entire recreate() function in case that full screen exclusive settings are changed. In addition, this adds button press based input events which detects the full screen exclusive mode, and tells if to trigger the recreate function. Furthermore, the init swapchain functions are modified per changes and configurations added to the header.

JeremyG2101 committed:
This is to introduce the Windows platform macro detection and related. If window platform is detected, then apply the full screen exclusive extensions and related functions.

gpx1000 committed:
add ninja-build to the Dockerfile.

JeremyG2101 committed:
This adds the first draft of README.md, meantime, organizing the snythax and format check

JeremyG2101 committed:
Finalzie the annotations, syntax and Clang format

JeremyG2101 committed:
Quick Save: This is a working verison, however, please DO NOT review it yet, for its completely messy, totally not human-readable. I will organize the code and add a group of logics.

JeremyG2101 committed:
This is the finalized version of Full Screen Exclusive sample. Where it organizes all functions and variables related to the extension; revises the Tutorial, and fixes several issues from the copyright.

JeremyG2101 committed:
Simply fixed some syntax issues from the Tutorial (i.e., README.md)

JeremyG2101 committed:
Revised some spelling and verb issues from README.md

gpx1000 committed:
Merged master into Full_Screen_Exclusive_Sample

JeremyG2101 committed:
Rerolling all framework related changes back to their original

JeremyG2101 committed:
fixing the windows command related issue

gpx1000 committed:
revert version of glslang

JeremyG2101 committed:
Fixing Windows platform detection using macro only

JeremyG2101 committed:
Merge branch 'Full_Screen_Exclusive_Sample' of https://bitbucket.org/holochip/vulkan-samples into Full_Screen_Exclusive_Sample

JeremyG2101 committed:
fix isFullScreenExclusive boolean undefined outside a Windows platform issue

JeremyG2101 committed:
Fix Recreate() function, that it is only defined on Windows platform

JeremyG2101 committed:
Bug-fixing and syntax/warning fix corresponding to Steves comments.

JeremyG2101 committed:
Snytax and format correction in README.md

JeremyG2101 committed:
fixed several warnings

JeremyG2101 committed:
fixing minior issues

JeremyG2101 committed:
removing some input variables which are not being used in the functions.

gpx1000 committed:
Merge remote-tracking branch 'upstream/main' into upstream-main

JeremyG2101 committed:
fixed a wording issue

JeremyG2101 committed:
fixed a wording issue from README.md

gpx1000 committed:
Merge branch 'upstream-main' into vk_Full_Screen_Exclusive

# Conflicts:
#	samples/CMakeLists.txt

gpx1000 committed:
merge with main.

gpx1000 committed:
revert to main's versions of dependencies

gpx1000 committed:
revert spirv-cross

gpx1000 committed:
fix an opps

asuessenbach commented on samples/extensions/full_screen_exclusive/README.md:
````diff
@@ -0,0 +1,363 @@
+following ```Enums``` can be chosen per selection of the display mode, Where:
+
+1. ```VK_FULL_SCREEN_EXCLUSIVE_DEFAULT_EXT```
+2. ```VK_FULL_SCREEN_EXCLUSIVE_DESALLOWED_EXT```
````
VK_FULL_SCREEN_EXCLUSIVE_D**I**SALLOWED_EXT

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.h:
```diff
@@ -0,0 +1,134 @@
+#include "core/instance.h"
+#include "platform/application.h"
+
+class FullScreenExclusive : public vkb::Application
```
Is there a reason, you're deriving your sample from vkb::Application and not from ApiVulkanSample or at least vkb::VulkanSample?

JeremyG2101 replied:
Ah yes. Reason for me to use vkb::Application as its base class is that, I can very easily get access to the swapchain, swapchain images, and image views (i.e., "swapchain-buffer"). The high-level access requires some "walk around" to access these variables (i.e., they were all defined inside "render-content" class). By directly using vkb::Application as a base class, it saves a lot of work, and it is more straightforward to showcase the fullscreen exclusive features.

asuessenbach reviewed (commented):
Although I'm not a native speaker, I think the readme.md needs some substantial word-smithing.
It probably should include the actual extension name `VK_EXT_full_screen_exclusive` used here at least once.

SaschaWillems commented:
I second @asuessenbach opinion on the readme. It's a bit hard to read. I think that's at least partially caused by too many words being put into tick marks:

![image](https://user-images.githubusercontent.com/10283127/216809824-5720f85d-2c62-46df-805d-0c74194f8fa5.png)

That makes it a bit hard to read. Using (code) tick marks to highlight code parts or extension names is fine, but I wouldn't use it for regular words.

JeremyG2101 committed:
changes wording, and reduces (cpp) tick marks

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1201 @@
+	instance_info.ppEnabledLayerNames     = requested_validation_layers.data();
+
+#if defined(VKB_DEBUG) || defined(VKB_VALIDATION_LAYERS)
+	VkDebugReportCallbackCreateInfoEXT debug_report_create_info = {VK_STRUCTURE_TYPE_DEBUG_REPORT_CREATE_INFO_EXT};
```
You might consider to use `VK_EXT_debug_utils` functionality here, instead of `VK_EXT_debug_report`, which is kind of deprecated.

SaschaWillems replied:
I second that. Actually this part crashes for me when one of the defines is set:

![image](https://user-images.githubusercontent.com/10283127/221171328-5c01707b-1854-4295-bfdf-cb012ee5e3fd.png)

`vkCreateDebugReportCallbackEXT` is NULL for me in that case.

asuessenbach commented on samples/extensions/full_screen_exclusive/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
```
As this should be a windows-only sample, you might make it windows-only by adding `if (WIN32)` here.

JeremyG2101 replied:
> As this should be a windows-only sample, you might make it windows-only by adding `if (WIN32)` here.

I actually took the conservative approach that, 1) I ensured that this sample runs on all platforms, 2) if it runs on a Windows platform, then it will showcase the fullscreen exclusive extension feature, otherwise 3) it simply is a "hello_triangle".
Which, I also mentioned such, in the introduction section in its README file.

asuessenbach replied:
I don't see, why it should be worth to have yet another hello_triangle sample named full_screen_exclusive for non-windows platforms.
I think, this sample doesn't give anything to non-windows users, so it shouldn't even be available on non-windows platforms.

JeremyG2101 replied:
> ** [asuessenbach](/asuessenbach)**

Okay, I will confirm this with Steve, and update it accordingly. Meantime, all the other changes are made corresponding to your comments.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1201 @@
+	active_instance_extensions.push_back(VK_KHR_PORTABILITY_ENUMERATION_EXTENSION_NAME);
+#endif
+
+#if defined(VK_USE_PLATFORM_ANDROID_KHR)
```
As this is a windows-only sample, all the other platforms don't need to be handled here.
Maybe just something like
```
#if defined(VK_USE_PLATFORM_WIN32_KHR)
...
#else
LOGE(...);
#endif
```

JeremyG2101 replied:
> As this is a windows-only sample, all the other platforms don't need to be handled here. Maybe just something like
> 
> ```
> #if defined(VK_USE_PLATFORM_WIN32_KHR)
> ...
> #else
> LOGE(...);
> #endif
> ```
Same reason addressed in the above comment section.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1201 @@
+
+	// If application is running on a Windows platform, then the following device extension is also needed:
+#if defined(VK_USE_PLATFORM_WIN32_KHR)
+	active_device_extensions.push_back(VK_EXT_FULL_SCREEN_EXCLUSIVE_EXTENSION_NAME);
```
You did not check the availability of `VK_EXT_FULL_SCREEN_EXCLUSIVE_EXTENSION_NAME`!
Maybe you can just pass it into this function via the `required_device_extensions`.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1201 @@
+			swapchain_size = surface_properties.currentExtent;
+		}
+	}
+#else
```
No need to handle non-win32, if it's only in win32 (controlled by its CMakeLists.txt).

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1201 @@
+		std::vector<VkQueueFamilyProperties> queue_family_properties(queue_family_count);
+		vkGetPhysicalDeviceQueueFamilyProperties(context.gpu, &queue_family_count, queue_family_properties.data());
+
+		for (int32_t j = 0; j < queue_family_count; j++)
```
warning C4018: '<': signed/unsigned mismatch

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1201 @@
+	if (!context.surface)
+		throw std::runtime_error("Failed to create window surface.");
+
+	init_device({"VK_KHR_swapchain"});
```
Maybe better: `VK_KHR_SWAPCHAIN_EXTENSION_NAME`

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.h:
```diff
@@ -0,0 +1,134 @@
+	static VkShaderStageFlagBits find_shader_stage(const std::string &ext);
+	void                         init_instance(const std::vector<const char *> &required_instance_extensions, const std::vector<const char *> &required_validation_layers);
+	void                         init_device(const std::vector<const char *> &required_device_extensions);
+	static void                  init_per_frame(Context &input_context, PerFrame &per_frame);
```
Why do some functions get the Context as input, and other do not? It's always the member variable.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.h:
```diff
@@ -0,0 +1,134 @@
+	static VkResult              acquire_next_image(Context &input_context, uint32_t *image);
+	static void                  render_triangle(Context &input_context, uint32_t swapchain_index);
+	static VkResult              present_image(Context &input_context, uint32_t index);
+	static void                  init_frame_buffers(Context &input_context);
```
Why do you rename `init_framebuffers` to `init_frame_buffers`?
There's nothing like buffers of a frame, there are just framebuffers.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.h:
```diff
@@ -0,0 +1,134 @@
+	static void                  render_triangle(Context &input_context, uint32_t swapchain_index);
+	static VkResult              present_image(Context &input_context, uint32_t index);
+	static void                  init_frame_buffers(Context &input_context);
+	static void                  teardown_frame_buffers(Context &input_context);
```
As above.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1201 @@
+	return true;
+}
+
+VkExtent2D FullScreenExclusive::update_current_maxImageExtent() const
```
Mixed camel- and snake-case looks odd. Should probably be `update_current_max_image_extent()`.
Moreover, this function does not update anything, it gets something. So maybe better `get_current_max_image_extent()`.
And, finally, this function is called just in init_swapchain. And at the top of that function you already get the surface capabilities. So, no need to have this function at all.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1201 @@
+
+		if (key_button.get_action() == vkb::KeyAction::Down)
+		{
+			bool isRecreate = false;
```
Maybe a better variable name: `needs_recreate`

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1201 @@
+	{
+		isWindowed = false;
+
+		GetWindowPlacement(HWND_applicationWindow, &wpc);
```
Can you get the `HWND_style` and `HWND_style_previous` just get in `initialize_windows`?

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1201 @@
+		GetWindowPlacement(HWND_applicationWindow, &wpc);
+		if (HWND_style == 0)
+			HWND_style = GetWindowLong(HWND_applicationWindow, GWL_STYLE);
+		if (HWND_style_previous == 0)
```
What makes `HWND_style_previous` previous? Shouldn't it be named `HWND_extended_style`, or so?

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1201 @@
+		long HWND_newStyle_previous = HWND_style_previous;
+		HWND_newStyle_previous &= ~WS_EX_WINDOWEDGE;
+
+		SetWindowLong(HWND_applicationWindow, GWL_STYLE, HWND_newStyle | static_cast<long>(WS_POPUP));
```
Why do you prepare the styles above, and add some bits here? Should probably be done in one block, for each of the styles.

JeremyG2101 committed:
README revised.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1201 @@
+		if (isFullScreenExclusive)
+		{
+			VkResult result = vkAcquireFullScreenExclusiveModeEXT(context.device, context.swapchain);
+			if (result == VK_SUCCESS)
```
The actual error code in case of failure would probably be of interest.
Maybe use `vkb::to_string(VkResult)` from `common/strings.h` here?

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1201 @@
+		{
+			VkResult result = vkAcquireFullScreenExclusiveModeEXT(context.device, context.swapchain);
+			if (result == VK_SUCCESS)
+			{
```
`vkReleaseFullScreenExclusiveModeEXT` is not used anywhere in this sample. Could you elaborate on why?

JeremyG2101 committed:
An update corresponding to the comments from asuessenbach: 1) the term "frame_buff" is now replaced as "framebuffer"; 2) all functions that had "context" as their input are now revised and such input is removed; 3) a static-cast from uint32_t to int32_t issue is fixed; 4) Syntax change using "VK_KHR_SWAPCHAIN_EXTENSION_NAME" everywhere only; 5) VK_STRUCTURE_TYPE_DEBUG_REPORT_CREATE_INFO_EXT is now replaced by VK_STRUCTURE_TYPE_DEBUG_UTILS_CREATE_INFO_EXT; 6) syntax change for name convension issues, and they are now unified; 7) Window placement variable is now registered only once inside the init_windows(); 8) All HWND styles are defined in one block; 9) added device extension availiablity check to: VK_EXT_FULL_SCREEN_EXCLUSIVE_EXTENSION_NAME; 10) added vkReleaseFullScreenExclusiveModeEXT function; 11) README revised per above changes.

JeremyG2101 committed:
Forgot to run Clang check........

gpx1000 committed:
use cmake to test for windows and below windows 11.  Don't list the sample unless it's a candidate version.  Remove system checks from the code directly.

gpx1000 committed:
Merge branch 'Full_Screen_Exclusive_Sample' into vk_Full_Screen_Exclusive

JeremyG2101 committed:
README.md has been revised based on the change made to use Windows 10 platform specifically.

gpx1000 committed:
small README.md corrections.

gpx1000 committed:
Merge remote-tracking branch 'downstream/Full_Screen_Exclusive_Sample' into vk_Full_Screen_Exclusive

SaschaWillems commented:
Aside from the crash when enabling validation (see above), the sample works fine for me.

One thing that's odd though is that the sample does start windowed. You have to lookup key codes to switch modes in the code first.

If the sample would have a UI with key codes visible at runtime and maybe information about the current state, that would make it easier to use.

SaschaWillems commented on samples/extensions/full_screen_exclusive/README.md:
```diff
@@ -0,0 +1,317 @@
+
+# Full Screen Exclusive
+
+## Overview
```
Maybe it's because I'm not a native (English) speaker, but this paragraph is kinda hard to read. Can we simplify the sentences here a little bit? From what I understand from this paragraph, the extension is only required to get exclusive full screen on Windows versions prior to 11?

SaschaWillems commented on samples/extensions/full_screen_exclusive/README.md:
```diff
@@ -0,0 +1,317 @@
+
+## Introduction
+
+This sample provides a detailed procedure to activate full screen exclusive mode on Windows 10 applications. Users can
```
This chapter refers to Windows 10. Does this also apply to older Windows 10 versions? If so, maybe update the text to be more in line with the first paragraph?

SaschaWillems commented:
One thing I'm missing from the readme: Why do people actually want to use this extension?

gpx1000 committed:
reword the tutorial.  Rely more upon showing in code than reproducing the code and cleanup the language.

SaschaWillems commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1172 @@
+	volkLoadInstance(context.instance);
+
+#if defined(VKB_DEBUG) || defined(VKB_VALIDATION_LAYERS)
+	VK_CHECK(vkCreateDebugReportCallbackEXT(context.instance, &debug_report_create_info, nullptr, &context.debug_callback));
```
This crashes for me on a secondary Workstation that doesn't have all the developement stuff setup. `vkCreateDebugReportCallbackEXT` is zero, and I think the problem might be related to how the preprocessor define are checked.

Validation layers (line 175) only check for `VKB_VALIDATION_LAYERS`, but the callback creation also has on or clause with `VKB_DEBUG`:

```cpp
#ifdef VKB_VALIDATION_LAYERS
	// Determine the optimal validation layers to enable that are necessary for useful debugging
	std::vector<const char *> optimal_validation_layers = vkb::get_optimal_validation_layers(supported_validation_layers);
	requested_validation_layers.insert(requested_validation_layers.end(), optimal_validation_layers.begin(), optimal_validation_layers.end());
#endif
...
#if defined(VKB_DEBUG) || defined(VKB_VALIDATION_LAYERS)
	VK_CHECK(vkCreateDebugReportCallbackEXT(context.instance, &debug_report_create_info, nullptr, &context.debug_callback));
#endif
```

On this system, where I did a clean checkout and did not change any settings this results in the aforementioned crash. Haven't been able to test on my personal desktop system yet.

All preprocessor defines should check for the same condition.

gpx1000 replied:
great catch.  All defines now agree.

gpx1000 committed:
fix the crash bug that Sascha caught.

Ensure that validation layers creation and extension insertion agree with each other.

SaschaWillems commented:
Almost ;)

Took me a while to realize this myself, but this doesn't work:

```cpp
#ifdef defined(VKB_DEBUG) || VKB_VALIDATION_LAYERS
	// Determine the optimal validation layers to enable that are necessary for useful debugging
	std::vector<const char *> optimal_validation_layers = vkb::get_optimal_validation_layers(supported_validation_layers);
	requested_validation_layers.insert(requested_validation_layers.end(), optimal_validation_layers.begin(), optimal_validation_layers.end());
#endif
```

Should be this instead:
```cpp
#if defined(VKB_DEBUG) || defined(VKB_VALIDATION_LAYERS)
	// Determine the optimal validation layers to enable that are necessary for useful debugging
	std::vector<const char *> optimal_validation_layers = vkb::get_optimal_validation_layers(supported_validation_layers);
	requested_validation_layers.insert(requested_validation_layers.end(), optimal_validation_layers.begin(), optimal_validation_layers.end());
#endif
```

SaschaWillems commented:
But even with the correct define, I still get `vkCreateDebugReportCallbackEXT` as zero. Can anyone else check this with a clean install?

In the framework we have a toggle that checks for debug utils. Maybe we need that in this sample too?

Edit: I think this is the culprit:

```cpp
#if defined(VKB_DEBUG) || defined(VKB_VALIDATION_LAYERS)
	active_instance_extensions.push_back(VK_EXT_DEBUG_UTILS_EXTENSION_NAME);
#endif
```

The sample enables the debug utils extension, but then uses `vkCreateDebugReportCallbackEXT` (from `VK_EXT_debug_report` ) instead of `vkCreateDebugUtilsMessengerEXT`.

gpx1000 committed:
use debug_utils instead of debug_report.  This change is probably needed in other samples too.

gpx1000 commented:
I don't have a windows computer near me that can run this sample.  So; I did this change blind.  Could you check to ensure I didn't do anything too silly.
This is probably something we should change in other samples too.

SaschaWillems commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -144,7 +142,7 @@ void FullScreenExclusive::init_instance(const std::vector<const char *> &require
 
 #if defined(VKB_DEBUG) || defined(VKB_VALIDATION_LAYERS)
-	active_instance_extensions.push_back(VK_EXT_DEBUG_UTILS_EXTENSION_NAME);
+	active_instance_extensions.push_back(VK_EXT_DEBUG_REPORT_EXTENSION_NAME);
```
Should be `VK_EXT_DEBUG_UTILS_EXTENSION_NAME`. With that, and the other changes this works fine in my office system too.

gpx1000 replied:
perfect; thanks for checking.

gpx1000 committed:
request the correct debug utils extension.

SaschaWillems commented:
On an Intel UHD Graphics 630, pressing F3 crashes the sample with `vkCreateSwapchainKHR` returning ` ERROR_INITIALIZATION_FAILED`. Not sure how to handle this though. Maybe we could catch that error and show a message that tells "this mode doesn't work with the current setup"?

gpx1000 committed:
F3 seems to crash it.  Use the suggested language when a crash happens.

gpx1000 commented:
Search or Find shortcut?  I had to lookup what F3 does.  Good to know.  I put in a check and an abort with your suggested error string.  Lemme know if you think it needs something else.  I'm not certain there's a good solution for that usecase though.

gpx1000 committed:
make clang-format happy

SaschaWillems reviewed (dismissed):
Thanks a lot for fixing this so fast 👍 

The sample now works as expected on my system with an NV RTX 2060 and another one with an Intel iGPU.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1177 @@
+	active_instance_extensions.push_back(VK_KHR_PORTABILITY_ENUMERATION_EXTENSION_NAME);
+#endif
+
+	LOGI("Windows Platform Detected, isWin32 set to be: true")
```
This log text seems to be a bit outdated.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1177 @@
+	// Get the HWND application window handle and stores it in the class variable
+	HWND_application_window = GetActiveWindow();
+	// Initialize full screen exclusive related variables since the application is now running on a windows platform
+	initialize_windows();        // this can be only called after GetActiveWindow(), or as long as HWND_application_window represent the correct handle!
```
Maybe move that call to `GetActiveWindow` to the top of `initialize_windows`, instead of commenting the need to call it first?

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1177 @@
+	std::vector<VkPhysicalDevice> gpus(gpu_count);
+	VK_CHECK(vkEnumeratePhysicalDevices(context.instance, &gpu_count, gpus.data()));
+
+	for (size_t i = 0; i < gpu_count && (context.graphics_queue_index < 0); i++)
```
Why is `i` a `size_t`, while `gpu_count` is an `uint32_t`?

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1177 @@
+		std::vector<VkQueueFamilyProperties> queue_family_properties(queue_family_count);
+		vkGetPhysicalDeviceQueueFamilyProperties(context.gpu, &queue_family_count, queue_family_properties.data());
+
+		for (int32_t j = 0; j < static_cast<int32_t>(queue_family_count); j++)
```
Why is `j` an `int32_t`, while `queue_family_count` is an `uint32_t`?

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1177 @@
+
+	if (context.graphics_queue_index < 0)
+	{
+		LOGE("Did not find suitable queue which supports graphics, compute and presentation.")
```
You don't look for a compute queue here.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1177 @@
+		throw std::runtime_error("Required device extensions are missing, will try without.");
+	}
+
+	std::vector<const char *> active_device_extensions = required_device_extensions;
```
Why do you copy the required extensions here?

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1177 @@
+	std::vector<VkSurfaceFormatKHR> formats(format_count);
+	vkGetPhysicalDeviceSurfaceFormatsKHR(context.gpu, context.surface, &format_count, formats.data());
+
+	VkSurfaceFormatKHR format;
```
Maybe change the format handling to how it's currently done in hello_triangle.

gpx1000 replied:
What are you asking for here?  The hello_triangle code in this function is fairly equivalent.  Are you looking for something specific?

asuessenbach replied:
Sascha had reworked that part lately (#622). I think, it would be good to adjust that part here as well

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1177 @@
+
+	if (is_full_screen_exclusive)
+	{
+		swapchain_size = get_current_max_image_extent();
```
No need to call `get_current_max_image_extent`, you can just use `surface_properties.maxImageExtent`.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1177 @@
+	VkSwapchainCreateInfoKHR info{};        // initialize the swapchain create info without adding pNext info
+	info.sType = VK_STRUCTURE_TYPE_SWAPCHAIN_CREATE_INFO_KHR;
+
+	// if this application is running on a Windows platform, then apply the full screen exclusive EXT, otherwise, nullptr.
```
The comment seems to be a bit outdated.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1177 @@
+	// This is to sync the platform with the class variable, so one may gain direct access to it
+	this->platform = &platform;
+
+	if (!context.surface)
```
Maybe move this check to right after surface creation call?

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1177 @@
+	if (!context.surface)
+		throw std::runtime_error("Failed to create window surface.");
+
+	// If application is running on a Windows platform, then the following TWO device extension is also needed:
```
"If application" -> "As application" ?
"is needed" -> "are needed"

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1177 @@
+
+void FullScreenExclusive::input_event(const vkb::InputEvent &input_event)
+{
+	// if application is running on a Windows platform:
```
Outdated comment?

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1177 @@
+
+			switch (key_button.get_code())
+			{
+				case vkb::KeyCode::F1:        // FullscreenExclusiveEXT = Disallowed
```
There definitely needs to be some output or UI telling the user about the supported keys!

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1177 @@
+							}
+						}
+
+						is_full_screen_exclusive = false;
```
Maybe move this into the body of the if-clause right above?

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1177 @@
+							}
+						}
+
+						is_full_screen_exclusive = false;
```
As above.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1177 @@
+		if (is_full_screen_exclusive)
+		{
+			VkResult result = vkAcquireFullScreenExclusiveModeEXT(context.device, context.swapchain);
+			LOGI("vkAcquireFullScreenExclusiveModeEXT: " + vkb::to_string(result))        // it would be necessary to learn the result on an unsuccessful attempt
```
You should only call `vkb::to_string(result)` if the result is not `VK_SUCCESS`.
Otherwise, you always get "UNKNOWN_ERROR" as output!
Or extend `vkb::to_string` to also handle that error code.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.h:
```diff
@@ -0,0 +1,130 @@
+		Fullscreen
+	};
+
+	struct SwapchainDimensions
```
Why do you introduce this struct?
You could just as well hold a `VkExtent2D` and a `VkFormat` in the Context struct.
At least the name of this struct is not appropriate, it holds more than just the dimensions.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.h:
```diff
@@ -0,0 +1,130 @@
+		VkSurfaceKHR               surface              = VK_NULL_HANDLE;
+		int32_t                    graphics_queue_index = -1;
+		std::vector<VkImageView>   swapchain_image_views{};
+		std::vector<VkFramebuffer> swapchain_frame_buffers{};
```
Maybe `swapchain_frame_buffers` -> `swapchain_framebuffers` ?

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.h:
```diff
@@ -0,0 +1,130 @@
+	LONG                                     HWND_style          = 0;                                            // current Hwnd style
+	LONG                                     HWND_extended_style = 0;                                            // previous Hwnd style
+	VkSurfaceFullScreenExclusiveInfoEXT      surface_full_screen_exclusive_info_EXT{};                           // it can be created locally, however, it is a good reminder that they are declared here as a class variable
+	VkSurfaceFullScreenExclusiveWin32InfoEXT surface_full_screen_exclusive_Win32_info_EXT{};                     // if using DirectX, then this variable has to be created and attach to the pNext of a VkSurfaceFullScreenExclusiveInfoEXT value
```
What do you mean with "If using DirectX..." ??

gpx1000 committed:
address most of the comments.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1172 @@
+		if (surface_properties.currentExtent.width == 0xFFFFFFFF)
+		{
+			swapchain_size.width  = context.swapchain_dimensions.width;
+			swapchain_size.height = context.swapchain_dimensions.height;
```
Simply `swapchain_size = context.swapchain_dimensions;` ?

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1172 @@
+	info.imageFormat        = format.format;
+	info.imageColorSpace    = format.colorSpace;
+	info.imageExtent.width  = swapchain_size.width;
+	info.imageExtent.height = swapchain_size.height;
```
Maybe `info.imageExtent = swapchain_size;` ?

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1172 @@
+		vkDestroySwapchainKHR(context.device, old_swapchain, nullptr);
+	}
+
+	context.swapchain_dimensions = {swapchain_size.width, swapchain_size.height };
```
Maybe `context.swapchain_dimensions = swapchain_size;` ?

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1172 @@
+	rp_begin.renderPass               = context.render_pass;
+	rp_begin.framebuffer              = framebuffer;
+	rp_begin.renderArea.extent.width  = context.swapchain_dimensions.width;
+	rp_begin.renderArea.extent.height = context.swapchain_dimensions.height;
```
Maybe rp_begin.renderArea.extent = context.swapchain_dimensions; ?

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1172 @@
+
+	VkRect2D scissor{};
+	scissor.extent.width  = context.swapchain_dimensions.width;
+	scissor.extent.height = context.swapchain_dimensions.height;
```
Maybe `scissor.extent = context.swapchain_dimensions;` ?

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1172 @@
+		throw std::runtime_error("Failed to create window surface.");
+	auto &extent                        = platform.get_window().get_extent();
+	context.swapchain_dimensions.width  = extent.width;
+	context.swapchain_dimensions.height = extent.height;
```
Maybe `context.swapchain_dimensions = extent;` ?

gpx1000 replied:
It's probably cleaner to not set the Window::Extent to be a type that it technically is not.  Although there is something to be said for our window class to use vkextent.

asuessenbach replied:
I think, I don't get what you mean, here.

gpx1000 replied:
Just meant that these are two different types.  There is no operator define in either structure that would allow assignment.

asuessenbach replied:
I see... you're right.

asuessenbach commented on samples/extensions/full_screen_exclusive/full_screen_exclusive.cpp:
```diff
@@ -0,0 +1,1172 @@
+	VK_CHECK(vkGetPhysicalDeviceSurfaceCapabilitiesKHR(context.gpu, context.surface, &surface_properties));
+
+	if (surface_properties.currentExtent.width == context.swapchain_dimensions.width &&
+	    surface_properties.currentExtent.height == context.swapchain_dimensions.height)
```
Maybe surface_properties.curentExtent = context.swapchain_dimensions; ?

gpx1000 replied:
that would change the logic.  This is a comparrison not an assignment.

gpx1000 committed:
Use Sascha's format handling code.  This is probably needed in several other samples too.

gpx1000 committed:
Display to console how to control the application

gpx1000 committed:
Merge branch 'main' into vk_Full_Screen_Exclusive
